PyRDF/Proxy.py

```diff
@@ -62,16 +84,56 @@ def GetValue(self):
         Value of the current action node
             This is the value obtained after executing the
             current action node in the computational graph.
-
         """
-        if not self.action_node.value:  # If event-loop not triggered
+        if not self.proxied_node.value:  # If event-loop not triggered
             from . import current_backend
-            generator = CallableGenerator(self.action_node.get_head())
+            generator = CallableGenerator(self.proxied_node.get_head())
             current_backend.execute(generator)
 
-        return self.action_node.value
+        return self.proxied_node.value
 
     def _call_handler(self, *args, **kwargs):
         # Handles an operation call to the current action node
         # and returns result of the current action node.
         return getattr(self.GetValue(), self._cur_attr)(*args, **kwargs)
+
+
+class TransformationProxy(Proxy):
+    """
+    A proxy object to an instantiated node. Used as a controller of the user
+    references to the node itself. When the user deletes reference to a
+    node (e.g. assigning the same variable to another operation), the proxy
+    object will get destroyed by Python, thus flagging the node to be without
+    user references anymore.
+    """
+
+    def __getattr__(self, attr):
```

Comment:
Since we are not getting attributes on `Node`s anymore (because now we always interact with a `Proxy` from the user application), the logic of the `__getattr__` in `Node` should be moved to `Proxy`. And the same thing for `_call_handler`.
It will require some code refactoring to understand what is common to both subclasses of `Proxy` and what needs to be done only in a specific subclass.

Reply:
I believe that such a refactoring would defeat the purpose of a proxy. The proxy should just mirror the methods and attributes of the node and maybe add little modifications (like the `has_user_references` attribute). But then, if all operation got directly managed and returned by proxies, what would be the purpose of the node? Furthermore, to keep all the information about the computational graph (callables included) we would also need to send information about proxies to the distributed executors (possibly via pickling) and right now we are not allowing that

Reply:
It depends on what functionality is outsourced to the `Proxy`. See how in `_call_handler` of `Node` we are going back to creating a `Proxy`:
https://github.com/JavierCVilla/PyRDF/blob/fbf79aec546dd816c901cbcd397f677b6ed80f88/PyRDF/Node.py#L167
This is an indicator that it should be the proxy itself that creates the node and stores it in its belly.
See also this:
https://github.com/JavierCVilla/PyRDF/blob/fbf79aec546dd816c901cbcd397f677b6ed80f88/PyRDF/Node.py#L167
such check is repeated here:
https://github.com/JavierCVilla/PyRDF/blob/fbf79aec546dd816c901cbcd397f677b6ed80f88/PyRDF/Proxy.py#L127
Having the `__getattr__` and `_call_handler` functionality split in two classes is not ideal, I would just have it in the class where you get the attribute and where you actually do the call.

In the end, `Node` can just be a class that is serialized to the backend and contains all the information needed for the execution. `Proxy` can take care of the checks and creation of the corresponding `Node`.